lightning/src/util/ser_macros.rs

```diff
@@ -1,3 +1,104 @@
+macro_rules! encode_tlv {
+	($stream: expr, {$(($type: expr, $field: expr)),*}) => { {
+		use bitcoin::consensus::Encodable;
+		use bitcoin::consensus::encode::{Error, VarInt};
+		use util::ser::{WriterWriteAdaptor, LengthCalculatingWriter};
+		$(
+			VarInt($type).consensus_encode(WriterWriteAdaptor($stream))
+				.map_err(|e| if let Error::Io(ioe) = e { ioe } else { unreachable!() })?;
+			let mut len_calc = LengthCalculatingWriter(0);
+			$field.write(&mut len_calc)?;
+			VarInt(len_calc.0 as u64).consensus_encode(WriterWriteAdaptor($stream))
+				.map_err(|e| if let Error::Io(ioe) = e { ioe } else { unreachable!() })?;
+			$field.write($stream)?;
+		)*
+	} }
+}
+
+macro_rules! encode_varint_length_prefixed_tlv {
+	($stream: expr, {$(($type: expr, $field: expr)),*}) => { {
+		use bitcoin::consensus::Encodable;
+		use bitcoin::consensus::encode::{Error, VarInt};
+		use util::ser::{WriterWriteAdaptor, LengthCalculatingWriter};
+		let mut len = LengthCalculatingWriter(0);
+		{
+			$(
+				VarInt($type).consensus_encode(WriterWriteAdaptor(&mut len))
+					.map_err(|e| if let Error::Io(ioe) = e { ioe } else { unreachable!() })?;
+				let mut field_len = LengthCalculatingWriter(0);
+				$field.write(&mut field_len)?;
+				VarInt(field_len.0 as u64).consensus_encode(WriterWriteAdaptor(&mut len))
+					.map_err(|e| if let Error::Io(ioe) = e { ioe } else { unreachable!() })?;
+				len.0 += field_len.0;
+			)*
+		}
+
+		VarInt(len.0 as u64).consensus_encode(WriterWriteAdaptor($stream))
+			.map_err(|e| if let Error::Io(ioe) = e { ioe } else { unreachable!() })?;
+		encode_tlv!($stream, {
+			$(($type, $field)),*
+		});
+	} }
+}
+
+macro_rules! decode_tlv {
+	($stream: expr, {$(($reqtype: expr, $reqfield: ident)),*}, {$(($type: expr, $field: ident)),*}) => { {
+		use ln::msgs::DecodeError;
+		let mut last_seen_type: Option<u64> = None;
+		'tlv_read: loop {
+			use bitcoin::consensus::encode;
+			use util::ser;
+			use std;
+
+			let typ: encode::VarInt = match encode::Decodable::consensus_decode($stream) {
+				Err(encode::Error::Io(ref ioe)) if ioe.kind() == std::io::ErrorKind::UnexpectedEof
+					=> break 'tlv_read,
+				Err(encode::Error::Io(ioe)) => Err(DecodeError::from(ioe))?,
+				Err(_) => Err(DecodeError::InvalidValue)?,
+				Ok(t) => t,
+			};
+
+			match last_seen_type {
+				Some(t) if typ.0 <= t => {
+					Err(DecodeError::InvalidValue)?
+				},
+				_ => {},
+			}
+			$(if (last_seen_type.is_none() || last_seen_type.unwrap() < $reqtype) && typ.0 > $reqtype {
+				Err(DecodeError::InvalidValue)?
+			})*
```

Comment:
For someone coming across this code, I think having the comments from my example would go a long way in understanding that these two two blocks are for.

Reply:
Oops, totally, thats my bad.